data/evidence/store.go

```diff
@@ -0,0 +1,273 @@
+package evidence
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/Oneledger/protocol/serialize"
+	"github.com/Oneledger/protocol/storage"
+)
+
+type EvidenceStore struct {
+	state  *storage.State
+	prefix []byte
+	mux    sync.Mutex
+}
+
+func NewEvidenceStore(prefix string, state *storage.State) *EvidenceStore {
+	return &EvidenceStore{
+		state:  state,
+		prefix: storage.Prefix(prefix),
+	}
+}
+
+func (es *EvidenceStore) WithState(state *storage.State) *EvidenceStore {
+	es.state = state
+	return es
+}
+
+func (es *EvidenceStore) Get(key []byte) ([]byte, error) {
+	prefixKey := append(es.prefix, key...)
+
+	dat, err := es.state.Get(storage.StoreKey(prefixKey))
+	if err != nil {
+		return nil, err
+	}
+	return dat, nil
+}
+
+func (es *EvidenceStore) GetVersioned(key []byte, height int64, diff int64) []byte {
+	prefixKey := append(es.prefix, key...)
+
+	return es.state.GetVersioned(height-diff, storage.StoreKey(prefixKey))
+}
+
+func (es *EvidenceStore) Set(key []byte, value []byte) error {
+	prefixKey := append(es.prefix, key...)
+	err := es.state.Set(storage.StoreKey(prefixKey), value)
+	return err
+}
+
+func (es *EvidenceStore) delete(key storage.StoreKey) (bool, error) {
+	prefixed := append(es.prefix, key...)
+	ok, err := es.state.Delete(prefixed)
+	if !ok || err != nil {
+		return ok, err
+	}
+	return ok, nil
+}
+
+func (es *EvidenceStore) getSuspiciousValidatorKey(validatorAddress keys.Address) []byte {
+	key := []byte(fmt.Sprintf("_ssvk_%s", validatorAddress))
+	return key
+}
+
+func (es *EvidenceStore) getSuspiciousVL() []byte {
+	key := []byte(fmt.Sprintf("_svvl"))
+	return key
+}
+
+func (es *EvidenceStore) GetSuspiciousValidator(validatorAddress keys.Address, height int64, diff int64) (*LastValidatorHistory, error) {
+	var dat []byte
+	if height == 0 {
+		dat, _ = es.Get(es.getSuspiciousValidatorKey(validatorAddress))
+	} else {
+		dat = es.GetVersioned(es.getSuspiciousValidatorKey(validatorAddress), height, diff)
+	}
+	if len(dat) == 0 {
+		return nil, fmt.Errorf("Suspisious validator not found")
+	}
+	lvh := &LastValidatorHistory{}
+	err := serialize.GetSerializer(serialize.PERSISTENT).Deserialize(dat, lvh)
+	if err != nil {
+		return nil, err
+	}
+	return lvh, nil
+}
+
+func (es *EvidenceStore) IterateSuspiciousValidators(fn func(lvh *LastValidatorHistory) bool) (stopped bool) {
+	key := []byte(fmt.Sprintf("_ssvk_"))
+	prefixKey := append(es.prefix, key...)
+	return es.state.IterateRange(
+		prefixKey,
+		storage.Rangefix(string(prefixKey)),
+		true,
+		func(key, value []byte) bool {
+			lvh, err := (&LastValidatorHistory{}).FromBytes(value)
+			if err != nil {
+				fmt.Println("failed to deserialize susp validator")
+				return false
+			}
+			if !lvh.IsFrozen() {
```

Comment:
Iterating thru 'Suspicious' validators means iterating thru 'Frozen' validators?

Reply:
I think this was named so because of the key thats used in the iterator function .